File: Trie/2079-delete-duplicate-folders-in-system/delete-duplicate-folders-in-system.cpp
class Solution {
    struct TreeNode{
        string val;
        TreeNode(string s): val(s){}
        map<string, TreeNode*> next;
    };
    TreeNode* root;
    unordered_map<string, int> key2Count;
    unordered_map<string, int> key2Id;
    unordered_map<TreeNode*, string> node2Key;
    vector<vector<string>> res;
public:
    vector<vector<string>> deleteDuplicateFolder(vector<vector<string>>& paths) {
        root = new TreeNode("/");
        for(auto path : paths){
            TreeNode* node = root;
            for(auto x : path){
                if(node->next.find(x) == node->next.end()) node->next[x] = new TreeNode(x);
                node = node->next[x];
            }
        }

        getId(root);

        vector<string> track;
        DFS(root, track);

        return res;
    }
    int getId(TreeNode* node){
        if(node == NULL) return 0;
        string key;

        for(auto x : node->next){
            TreeNode* child = x.second;
            key += to_string(getId(child)) + "#" + child->val + "#";
        }

        node2Key[node] = key;
        key2Count[key] += 1;
        if(key2Count[key] == 1){
            key2Id[key] = key2Id.size() + 1;
        }

        return key2Id[key];
    }

    void DFS(TreeNode* node, vector<string>& track){
        
        string key = node2Key[node];
        if(key != "" && key2Count[key] >= 2){
            return;
        }

        if(node->val != "/"){
            track.push_back(node->val);
            res.push_back(track);
        }

        for(auto x : node->next){
            DFS(x.second, track);
        }

        if(node->val != "/"){
            track.pop_back();
        }
    }
};

File: Trie/2079-delete-duplicate-folders-in-system/delete-duplicate-folders-in-system.py
class TrieNode:
    __slots__ = 'son', 'name', 'deleted'

    def __init__(self):
        self.son = {}
        self.name = ''
        self.deleted = False


class Solution:
    def deleteDuplicateFolder(self, paths: List[List[str]]) -> List[List[str]]:
        root = TrieNode()
        for path in paths:
            cur = root
            for s in path:
                if s not in cur.son:
                    cur.son[s] = TrieNode()
                cur = cur.son[s]
                cur.name = s

        expr_to_node = {}

        def gen_expr(node: TrieNode) -> str:
            if not node.son:
                return node.name

            expr = sorted('(' + gen_expr(son) + ')' for son in node.son.values())
            sub_tree_expr = ''.join(expr)
            if sub_tree_expr in expr_to_node:
                expr_to_node[sub_tree_expr].deleted = True
                node.deleted = True 
            else:
                expr_to_node[sub_tree_expr] = node

            return node.name + sub_tree_expr

        for son in root.son.values():
            gen_expr(son)

        ans = []
        path = []

        def dfs(node: TrieNode) -> None:
            if node.deleted:
                return
            path.append(node.name)
            ans.append(path.copy())
            for child in node.son.values():
                dfs(child)
            path.pop()

        for son in root.son.values():
            dfs(son)

        return ans
